src/CommunityToolkit.Maui/Layouts/StateLayout/StateLayoutController.shared.cs

```diff
@@ -0,0 +1,326 @@
+/// StateLayout Controller
+/// </summary>
+public class StateLayoutController: IDisposable
+{
+	readonly WeakReference<Layout> layoutWeakReference;
+	bool layoutIsGrid;
+	LayoutState previousState = LayoutState.None;
+	IList<View> originalContent = Enumerable.Empty<View>().ToList();
+	CancellationTokenSource? animationTokenSource;
+
+	bool isDisposed;
+
+	/// <summary>
+	/// The StateViews defined in the StateLayout.
+	/// </summary>
+	public IList<StateView> StateViews { get; set; } = Enumerable.Empty<StateView>().ToList();
+
+	/// <summary>
+	/// StateLayout Controller constructor
+	/// </summary>
+	/// <param name="layout"></param>
+	public StateLayoutController(Layout layout)
+		=> layoutWeakReference = new WeakReference<Layout>(layout);
+	
+	/// <summary>
+	/// Display the default content.
+	/// </summary>
+	/// <param name="animate"></param>
+	public async void SwitchToContent(bool animate)
+	{
+		if (!layoutWeakReference.TryGetTarget(out var layout))
+		{
+			return;
+		}
+
+		var token = RebuildAnimationTokenSource(layout);
+
+		previousState = LayoutState.None;
+		await ChildrenFadeTo(layout, animate, true);
+
+		if (token.IsCancellationRequested)
+		{
+			return;
+		}
+
+		// Put the original content back in.
+		layout.Children.Clear();
+
+		foreach (var item in originalContent)
+		{
+			item.Opacity = animate ? 0 : 1;
+			layout.Children.Add(item);
+		}
+
+		await ChildrenFadeTo(layout, animate, false);
+	}
+
+	/// <summary>
+	/// Display the StateView template for the given custom state key.
+	/// </summary>
+	/// <param name="customState"></param>
+	/// <param name="animate"></param>
+	public void SwitchToTemplate(string customState, bool animate)
+		=> SwitchToTemplate(LayoutState.Custom, customState, animate);
+
+	/// <summary>
+	/// Display the StateView template for the given LayoutState option.
+	/// </summary>
+	/// <param name="state"></param>
+	/// <param name="customState"></param>
+	/// <param name="animate"></param>
+	/// <exception cref="ArgumentException"></exception>
+	public async void SwitchToTemplate(LayoutState state, string? customState, bool animate)
+	{
+		if (!layoutWeakReference.TryGetTarget(out var layout))
+		{
+			return;
+		}
+
+
+		var token = RebuildAnimationTokenSource(layout);
+
+		// Put the original content somewhere where we can restore it.
+		if (previousState == LayoutState.None)
+		{
+			originalContent = new List<View>();
+
+			foreach (var item in layout.Children)
+			{
+				originalContent.Add((View)item);
+			}
+
+		}
+
+		var view = GetViewForState(state, customState);
+
+		if (view != null)
+		{
+			previousState = state;
+
+			await ChildrenFadeTo(layout, animate, true);
+
+			if (token.IsCancellationRequested)
+			{
+				return;
+			}
+
+
+			layout.Children.Clear();
+
+			var repeatCount = GetRepeatCount(state, customState);
+			var template = GetTemplate(state, customState);
+
+			if (template != null)
+			{
+				// We have a template we can use.
+				var items = new List<int>();
```

Comment:
We already know the size of the `List`, so we can take advantage of that.
```suggestion
				var items = new List<int>(repeatCount);
```